---
title: "DEMO"
author: "JR"
date: "2024-09-25"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(IRanges)
library(dplyr)
library(tidyr)
library(tibble)
library(readr)
library(ggplot2)
library(purrr)
library(magrittr)
library(pheatmap)
library(textshape)
library(Rcpp)
library(DESeq2)
library(matrixStats)
library(biomaRt)
library(org.Mm.eg.db)
library(tinytex)
BiocManager::install("biomaRt") 
BiocManager::install("org.Mm.eg.db")
```

So far, we have done a lot of fundamental RNA sequencing analysis - great work!
Now let's compile this and compress it into one RMD file. This will be a great way to
(i) practice what we have learned, (ii) a useful resource for any future RNAseq you may want to do
and (iii) ALL the data we need to start making figures - our next section :)

# Goal: script our RNAseq analysis pipeline to find genes changing significant across time from dox exposure. 

# Importing counts and TPM values
```{r import counts and TPM from Salmon}
# laoding Salmon counts from NF_CORE run 3.14
counts_matrix <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/lessons/04_RNAseq_Dox/01_Mouse_dox_wt/good_class_RNAseq/pipeline_output/star_salmon/salmon.merged.gene_counts.tsv", header=TRUE, row.names=1)

counts_matrix <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/tess/MASTER_CLASS/data/salmon.merged.gene_counts.tsv", header=TRUE, row.names=1)



# preparing counts for DESEQ
# create g2s
g2s <- data.frame(
  gene_id = rownames(counts_matrix),
  gene_name = counts_matrix[, 1]
)
# removing gene_name column for analyses
counts_matrix <- counts_matrix[ , -1]
# Round counts to integer mode required for DESEQ2
counts_integer <- round(counts_matrix)

```

# Creating sample sheet for deseq (coldata) from counts file
```{r creating deseq sample sheet (coldata)}
# setting up data frame with cols from counts matrix
deseq_samples <- data.frame(
  sample_id = colnames(counts_matrix))
# creating time point and replicate cols from col names
# Split names
split_values <- strsplit(deseq_samples$sample_id, "_")
# graabbing time point and replicate info 
time_values <- sapply(split_values, function(x) x[[2]])
replicate_values <- sapply(split_values, function(x) x[[3]])
# placing time point and replicate info into dataframe
deseq_samples$time_point <- time_values
deseq_samples$replicate <- replicate_values
# IMPORTANT : Factor time point and replicate
deseq_samples$time_point <- factor(deseq_samples$time_point)
deseq_samples$replicate <- factor(deseq_samples$replicate)

```

# Running DESEQ2
```{r running deseq2}
# testing sample sheet and counts are arranged properly 
stopifnot(all(colnames(counts_integer) == rownames(deseq_samples$sample_id)))

# setting up dds
dds_time_point <- DESeqDataSetFromMatrix(countData = counts_integer,
                              colData = deseq_samples,
                              design = ~ time_point)
# Run time point model
dds_time_point <- DESeq(dds_time_point)

```

# Extracting results for each time point from DESEQ dds
```{r compile DESEQ2 results}

# setting up result names object for input to forloop
result_names <- resultsNames(dds_time_point)
results_names <- result_names[-1]

# Setting up dataframe
res_df <- data.frame("gene_id" = character(), 
                     "baseMean" = numeric(), 
                     "log2FoldChange" = numeric(), 
                     "lfcSE" = numeric(),
                     "stat" = numeric(),
                     "pvalue" = numeric(),
                     "padj" = numeric(),
                     "gene_name" = character(),
                     "result_name" = character())

# run for loop to retreive results from all time comparisons in dds
for(i in 1:length(results_names)) {
  results_name <- results_names[i]
  res <- results(dds_time_point, name = results_name)
  tmp_res_df <- res %>% 
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    merge(g2s) %>%
    mutate(result_name = results_name)
  res_df <- bind_rows(res_df, tmp_res_df)
  
}

```

# Filtering to significant results P <0.05 and LFC abs(1)
```{r filtering DESEQ results to LFC > abs(1)& P <0.05}

filtered_res_df <- res_df %>%
  filter(padj < 0.05, abs(log2FoldChange) > 1)

unique_sig_genes <- unique(filtered_res_df$gene_id)

unique_sig_genes <- unique(filtered_res_df$gene_id)
unique_sig_genes_df <- data.frame(gene_id = unique_sig_genes)
```

# Nice only 100 lines of code to go from counts table to DESEQ results !
Now your future self can go from raw sequencing files to NF_CORE pipeline to DESEQ analysis.
That should be pretty handy if you ever do RNAseq in the future.

# NOW LET'S SAVE ALL OUR WORK FOR MAKING FIGURES !
Since all this data is from counts, let's make a results directory
in our results directory :) called 'count_results_DESEQ'

```{r saving DESEQ results}

# save(unique_sig_genes, counts_integer, g2s, res_df, filtered_res_df, deseq_samples, file = "/scratch/Shares/rinnclass/MASTER_CLASS/lessons/06_Differential_expression_analyses/results/counts_results_DESEQ/DESEQ_results.rdata")

save(unique_sig_genes, counts_integer, g2s, res_df, filtered_res_df, deseq_samples, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/Mich4366/MASTER_CLASS/lessons/06_Differential_expression_analyses/My_Project/results/DESEQ_results.rdata")

# load("results/counts_results_DESEQ/DESEQ_results.rdata")
# load ("My_Project/results/DESEQ_results.rdata")
```

# As a final step let's archive the average and standard deviation of our TPM values
After this last step we will have all the info we need for plotting in lv 7 !

# Import TPM file from Salmon
```{r import counts and TPM from Salmon}
# laoding and preparing TPM file from Salmon

TPM <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/lessons/04_RNAseq_Dox/01_Mouse_dox_wt/data/results/star_salmon/salmon.merged.gene_tpm.tsv", header=TRUE, row.names=1)

TPM <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/tess/MASTER_CLASS/data/salmon.merged.gene_tpm.tsv", header=TRUE, row.names=1)

TPM <- TPM %>%
     select(-gene_name)

# Filtering for TPM of at least 1 across samples.
TPM_filtered <- TPM[rowSums(TPM) > 1, ]
# Checking
any(rowSums(TPM_filtered) < 1)
any(rowSums(TPM) < 1)

```

# Calculating mean and sd for each gene in each time point (replicate mean and sd)
```{r making table of mean and sd TPM value for each gene for each time point}
#install.packages("matrixStats")
# library(matrixStats)

# Time and replicate values
time_points <- c("0", "12", "24", "48", "96")
replicates <- c("_1", "_2", "_3")

# initialize list for results
average_and_stddev_values <- list()
# now each object into the for loop is "tp"
for (tp in time_points) {
  cols <- grep(paste0("WT_", tp, "_"), colnames(TPM_filtered))
  avg <- rowMeans(TPM_filtered[, cols])
  std_dev <- apply(TPM_filtered[, cols], 1, sd)
  std_dev <- data.frame(std_dev)
  combined <- cbind(avg, std_dev)
  average_and_stddev_values <- c(average_and_stddev_values, list(combined))
}
# Convert the list to a data frame
average_and_stddev_values <- do.call(cbind, average_and_stddev_values)
# Add column names for the time points
colnames(average_and_stddev_values) <- paste0(rep(time_points, each = 2), c("_avg", "_sd"))
```

# Now let's save our TPM results as a seperate object
```{r saving TPM data}

# note I am adding in some other files we just saved, this is in case
# we just want to open up TPM data we still have G2s and Samplesheet.
# save(TPM, average_and_stddev_values, TPM_filtered, g2s, deseq_samples, file = "/scratch/Shares/rinnclass/MASTER_CLASS/lessons/06_Differential_expression_analyses/results/TPM_results/TPM_results.Rdata" )

save(TPM, average_and_stddev_values, TPM_filtered, g2s, deseq_samples, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/Mich4366/MASTER_CLASS/lessons/06_Differential_expression_analyses/My_Project/results/TPM_results.Rdata" )

# test
# load("results/TPM_results/TPM_results.Rdata")
# load ("My_Project/results/TPM_results.Rdata")


#Pulling out the genes of the brain from the data set 

#You can use biomaRt to fetch annotations for your Ensembl gene IDs. Here is how you can set it up:

# Set up a connection to the Ensembl database for Mus musculus (mouse)
ensembl = useMart("ensembl", dataset = "mmusculus_gene_ensembl") # big list of genes with descriptions

# Retrieve gene annotations using Ensembl gene IDs
gene_data <- getBM(attributes = c("ensembl_gene_id", "external_gene_name", "gene_biotype", "description"),mart = ensembl)

# Example list of brain-related keywords
brain_keywords <- c( #key words for finding which genes have to do with the brain
  "brain", "neuron", "cortex", "hippocampus",
  "amygdala", "thalamus", "hypothalamus", "cerebellum", "brainstem",
  "medulla", "pons", "midbrain", "frontal lobe", "parietal lobe",
  "temporal lobe", "occipital lobe", "white matter", "gray matter",
  "synapse", "axon", "dendrite", "glia", "astrocyte", "microglia",
  "oligodendrocyte", "myelin", "neurotransmitter", "dopamine",
  "serotonin", "GABA", "acetylcholine", "neuroplasticity",
  "neurogenesis", "electrophysiology", "fMRI", "EEG", "BOLD signal",
  "action potential", "resting potential", "ion channel"
)

# Filter the data for brain-related genes
brain_genes <- gene_data[grep(paste(brain_keywords, collapse = "|"), gene_data$description, ignore.case = TRUE), ] # list of brain genes with description

#Assuming you have a dataset (my_data) with Ensembl gene IDs in the "meta" form, you can merge this with your filtered brain_genes dataset:

#Remove ensembl portion of the gene id 
colnames(brain_genes) <- gsub("external_","",colnames(brain_genes)) # fixing a column name
colnames(brain_genes) <- gsub("ensembl_","",colnames(brain_genes))

filtered_res_df$gene_id <- gsub("\\..*", "", filtered_res_df$gene_id)
rownames(TPM_filtered) <- gsub("\\..*", "", rownames(TPM_filtered))

brain_genes_descriptions <- merge(brain_genes,filtered_res_df, by = "gene_name")
merged_with_replicates <- cbind(gene_id = rownames(TPM_filtered), TPM_filtered) # putting gene id as one of the columns

colnames(brain_genes_descriptions) <- gsub(".x","",colnames(brain_genes_descriptions)) # fixing a column name
brain_gene_replicates <- merge(brain_genes_descriptions, merged_with_replicates, by = "gene_id")
brain_gene_replicates <- brain_gene_replicates[,-c(2:12)]

```
# VERY NICE we have all our results organized and can start to make figures
# WE ARE READY FOR LEVEL 7 - Congrats on finishing Level 6 !!



###################
## SOME FIGURES ##
###################


```{R}
sig_genes <- as.data.frame(unique(filtered_res_df$gene_id))

names(sig_genes) <- "gene_id"

average_and_stddev_values <- rownames_to_column(average_and_stddev_values, "gene_id")

# one way to subset the data frame of avrg / sd TPM values
sig_gene_tpm <- inner_join(average_and_stddev_values, sig_genes, by = "gene_id")

# Same can be done by indexing
sig_gene_tpm_2 <- average_and_stddev_values[average_and_stddev_values$gene_id %in%
                                                sig_genes$gene_id, ]

# let's now merge in the gene names into this handy dataframe
sig_gene_tpm <- merge(sig_gene_tpm, g2s)

# Now we should organize so we can plot specific genes (not mix character and integer)
sig_gene_tpm <- sig_gene_tpm %>%
  select(-gene_id) 


# cool now we have the tpm and sd values to plot !
```


# Line plot from full TPM data frame.
Here we are going to do something more typical is plot the same line
plots above but using the replicate information. This way we are more 
flexible going to the full table with replicates. For example, we could 
calculate standard error instead of standard deviation this way.

# Logic flow:
(1) Have a set of genes of interest
(2) Reduce TPM table to those genes
(3) Pivot longer so all TPM data is in one column (genes have multiple rows)
(4) Group by and summarize to get mean and sd 
(5) Plot those values

```{r line plot from large table}

# (1) set of genes of interest or list.
genes_of_interest <- data.frame(gene = c("ENSMUSG00000094125.1", "ENSMUSG00000079247.2", "ENSMUSG00000112027.1", "ENSMUSG00000066632.3", "ENSMUSG00000000031.16"))

# (2) Reduce TPM file to these genes
df_plot_genes <- TPM_filtered[rownames(TPM_filtered) %in% genes_of_interest$gene, ]

# Moving rownames in to a column
df_plot_genes <- rownames_to_column(df_plot_genes, var = "gene_id")

# (3) Pivot longer (watch the changes !)
df_plot_genes <- df_plot_genes %>% 
  pivot_longer(cols = -gene_id, names_to = 'Time_Replicate', values_to = 'TPM')

 brain_gene_replicates_turned <- brain_gene_replicates %>% 
     pivot_longer(cols = -gene_id, names_to = 'Time_Replicate', values_to = 'TPM')

# Seperating sample name into time and replicate for plotting table
df_plot_genes <- df_plot_genes %>%
  separate(
    Time_Replicate,
    into = c('WT', 'Time', 'Replicate'),
    sep = '_',
    extra = "merge"
  ) %>%
  select(-WT)


 brain_gene_replicates_turned <- brain_gene_replicates_turned %>%
     separate(
         Time_Replicate,
       into = c('WT', 'Time', 'Replicate'),
         sep = '_',
        extra = "merge"
    ) %>%
   dplyr::select(-WT)

# Convert Time to numeric values
df_plot_genes$Time <- as.numeric(df_plot_genes$Time)

brain_gene_replicates_turned$Time <- as.numeric(brain_gene_replicates_turned$Time)

# (4) Group by Time and calculate mean and std of TPM
df_plot_genes <- df_plot_genes %>% 
  group_by(gene_id, Time) %>% 
  summarise(mean_TPM = mean(TPM), sd_TPM = sd(TPM))

 brain_gene_replicates_turned <- brain_gene_replicates_turned%>% 
    group_by(gene_id, Time) %>% 
    summarise(mean_TPM = mean(TPM), sd_TPM = sd(TPM))

# (5) Plot the line plot with error bars
ggplot(df_plot_genes, aes(x = Time, y = mean_TPM)) +
  geom_line() +
  geom_errorbar(aes(ymin = mean_TPM - sd_TPM, ymax = mean_TPM + sd_TPM), width = 0.2) +
  labs(x = 'Time (hours)', y = 'Average TPM') +
  ggtitle('Average TPM across Time Points') +
  facet_wrap(~ gene_id)

ggplot(brain_gene_replicates_turned, aes(x = Time, y = mean_TPM)) +
  geom_line() +
  geom_errorbar(aes(ymin = mean_TPM - sd_TPM, ymax = mean_TPM + sd_TPM), width = 0.2) +
  labs(x = 'Time (hours)', y = 'Average TPM') +
  ggtitle('Average TPM across Time Points') +
  facet_wrap(~ gene_id)

brain_gene_replicates_no_id <- brain_gene_replicates
 rownames(brain_gene_replicates_no_id) <- make.unique(as.character(brain_gene_replicates_no_id$gene_id))
 brain_gene_replicates_no_id$gene_id <- NULL  # remove the column

log_tpm_matrix <- log2(TPM_filtered + 1)

log_brain_gene_replicates_TPM <- log2(brain_gene_replicates_no_id+1)

 distance_matrix <- dist(t(log_brain_gene_replicates_TPM), method = "euclidean")
 hc_complete <- hclust(distance_matrix, method = "complete")
 plot(hc_complete, labels = colnames(log_brain_gene_replicates_TPM), main = "Global TPM Filtered Dendrogram ", sub = "sample relationship by time and replicate after dox treatment")

clusters <- cutree(hc_complete, k = 5)

print(clusters)

dend_hc_complete <- as.dendrogram(hc_complete)


view(dend_hc_complete)

# Reorder the dendrogram based on the clusters created above
dend_hc_complete <- reorder(dend_hc_complete, clusters)

# Plot the reordered dendrogram
plot(dend_hc_complete, main = "Global TPM Filtered Dendrogram (Ordered by Cluster)", sub = "Sample relationship by time and replicate after dox treatment")

pheatmap(log_brain_gene_replicates_TPM, 
         cluster_rows = TRUE,  # Cluster genes (rows)
         cluster_cols = TRUE,  # Cluster samples (columns)
        scale = "row",        # Scale the data across rows (genes) for better visualization
          show_rownames = FALSE, # Hide gene names for clarity
          show_colnames = TRUE,  # Show sample names
          main = "Heatmap of Log2-Transformed TPM Values for Sig Brain Genes")


custom_dist <- function(x) dist(x, method = "manhattan") # Replace "manhattan" with your preferred method
custom_hclust <- function(x) hclust(x, method = "ward.D2") # Replace "ward.D2" with your preferred clustering method

# Create a heatmap with custom distance and clustering functions
pheatmap(log_tpm_matrix, 
         cluster_rows = TRUE,  
         cluster_cols = TRUE,  
         scale = "row",        
         show_rownames = FALSE, 
         show_colnames = TRUE,  
         main = "Heatmap of Log2-Transformed TPM Values",
         color = colorRampPalette(c("blue", "white", "red"))(50),
         distfun = custom_dist,   
         hclustfun = custom_hclust 
)


#ler3 IGV no significant data 
#nsmf  IGV no significant
#dab2 IGV no significant 
#mal no significant
#Crmp1 no significant
#Slc6a13 no significant 
#Trpc6
# Myrf significant data seen 
# Htr5a
# Nnat some data seen

# This could be cleaned up by using gene_name instead -- fun exeRcise :)
```

#####HEATMAP#######

```{r load dds for mouse long time course DESEQ2}

# First Normalize counts (rlog) function in DESEQ2
# This basically is rank counts normalized to std error in replicates.
rlog_counts_mouse_long <- rlog(dds_time_point, blind = TRUE)
# library(DESeq2)
# Now we retrieve the values using the "assay" function that converts to rlog_counts)
rlog_counts_matrix_mouse_long <- assay(rlog_counts_mouse_long)
# which(is.na(rlog_counts_mouse_long))

breaks <- seq(-2, 2, length.out = 51)

# Filter to sig genes
sig_rlog_counts_matrix_mouse_long <- rlog_counts_matrix_mouse_long[rownames(rlog_counts_matrix_mouse_long) %in% unique_sig_genes, ]

# Plotting heatmap
pheatmap(sig_rlog_counts_matrix_mouse_long, 
         cluster_rows = TRUE,  
         cluster_cols = TRUE,  
         scale = "row",        
         show_rownames = FALSE, 
         show_colnames = TRUE,  
         main = "Heatmap of Sig genes Rlog-Transformed Values",
         breaks = breaks,
         color = colorRampPalette(c("blue", "white", "red"))(50) # Gradient from blue to red
         )



